test/permissions/remote.provider.get.js

```diff
@@ -0,0 +1,191 @@
+'use strict';
+
+const jstp = require('@metarhia/jstp');
+
+const { codes: errorCodes } = require('../../lib/errors');
+
+const test = (test, { categories, users, roles, data, permissions }) => {
+  const expectedErrorCode = errorCodes.INSUFFICIENT_PERMISSIONS;
+  const expectedError = new jstp.RemoteError(expectedErrorCode);
+
+  test.endAfterSubtests();
+
+  test.test(
+    'remoteProvider.get: check read permission, sufficient permission',
+    (test, { remoteProvider, login }) =>
+      login(users.S1.ReadOnlyUser, () =>
+        remoteProvider.get(categories.Role, (err, res) => {
+          if (err) {
+            test.bailout(err.stack);
+          }
+          test.assert(res);
+          test.end();
+        })
+      )
+  );
+
+  test.test(
+    'remoteProvider.get: check read permission, no userId',
+    (test, { remoteProvider }) =>
+      remoteProvider.get(categories.Role, (err, res) => {
+        test.isError(err, expectedError);
```

Comment:
Would it perhaps be better to extract this in a function
```js
const isError = (test, err, res) {
  test.strictSame(err.code, expectedErrorCode);
  test.assertNot(res);
  test.end();
})
```
and then `isError.bind(null, test)`?

Reply:
I thought about this too, however, in some tests there can be some differences when checking results in different methods in one test file. For example, in some cases I have to check result like this:
```javascript
test.assertNot(res);
```
and in some cases like this:
```javascript
test.assertNot(res[0]);
```
So would it be okay to use in some cases a custom binded `isError`  and in other cases the whole block with some minor changes?

Reply:
@lundibundi you have a point, this looks better in any case. WDYT about making the same for successful subtests? They have also the same structure, so can be extracted to a separate function.

Reply:
That's okay with me.

Also, you can try to use a more generic approach where you will define some cases as an array `[ { method, args, expectedResult, expectedError }, { ... } ]` and have some executor for those cases. (e.g. you can take a look at https://github.com/metarhia/common/blob/master/test/int64.js).